app/domain/services/create_course/service.rb

```diff
@@ -0,0 +1,39 @@
+class Services::CreateCourse::Service
+  def process(course_uuid:, ecosystem_uuid:)
+    return nil if course_uuid.nil? or ecosystem_uuid.nil?
+
+    start_time     = Time.now
+    start_time_str = start_time.utc.iso8601(6)
+
+    ecosystem = Ecosystem.find_by uuid: ecosystem_uuid
+
+    fail Errors::AppRequestValidationError.new("Ecosystem #{ecosystem_uuid} does not exist. Course cannot be created.") \
+      if ecosystem.nil?
+
+    values_str = %Q{
+      ( '#{course_uuid}',
+        '#{ecosystem_uuid}',
+        TIMESTAMP WITH TIME ZONE '#{start_time_str}',
+        TIMESTAMP WITH TIME ZONE '#{start_time_str}' )
+    }.gsub(/\n\s*/, ' ')
+
+    created_course_uuid = Course.transaction(isolation: :serializable) do
+      sql_inserted_course = %Q{
+        INSERT INTO courses
+        (uuid,ecosystem_uuid,created_at,updated_at)
+        VALUES #{values_str}
+        ON CONFLICT DO NOTHING
+        RETURNING uuid
+      }.gsub(/\n\s*/, ' ')
+
+      inserted_course = Course.connection.execute(sql_inserted_course)
+
+      inserted_course.collect{|hash| hash.fetch('uuid')}[0]
```

Comment:
I don't think this is actually idempotent; if the course had previously been inserted, `created_course_uuid` could be empty (I believe it should be plural, since a collection is returned).  A second query would need to be made to in that case to ensure really does exist, but that could also be inferred (in this case) from the fact that only one thing is being inserted and nothing was returned and nothing failed.